### PR TITLE
Re-implement VoxelFilter with unordered_set

### PR DESCRIPTION
**Pull request by gaschler**
Use hashing to filter voxels.
Handles arbitrary grid coordinates up to sizeof(int).

FIXES=#937

### CODE REVIEW

**Comment by gaschler** [comment]
I did a quick performance evaluation it also seems faster:
```
PointCloud CreatePointCloud() {
  std::mt19937 rng(1);
  std::uniform_real_distribution<float> distribution(-100, 100);
  PointCloud point_cloud;
  for (int i = 0; i < 1000; ++i) {
    const auto x = distribution(rng);
    const auto y = distribution(rng);
    const auto z = distribution(rng);
    point_cloud.emplace_back(x, y, z);
  }
  return point_cloud;
}

TEST(VoxelFilterTest, VoxelFilterMany) {
  PointCloud input = CreatePointCloud();
  for (int i = 0; i < 1000; ++i) {
    auto output = VoxelFilter(1.f).Filter(input);
  }
}

TEST(VoxelFilterTest, SimpleVoxelFilterMany) {
  PointCloud input = CreatePointCloud();
  for (int i = 0; i < 1000; ++i) {
    auto output = SimpleVoxelFilter(1.f).Filter(input);
  }
}
```

[ RUN      ] VoxelFilterTest.VoxelFilterMany
[       OK ] VoxelFilterTest.VoxelFilterMany (419 ms)
[ RUN      ] VoxelFilterTest.SimpleVoxelFilterMany
[       OK ] VoxelFilterTest.SimpleVoxelFilterMany (181 ms)

(SimpleVoxelFilter is the new one in this comparison, now replaces VoxelFilter.)

File: cartographer/sensor/voxel_filter.h [hunk 1]
```diff
@@ -31,10 +33,7 @@ namespace sensor {
 class VoxelFilter {
  public:
   // 'size' is the length of a voxel edge.
-  explicit VoxelFilter(float size);
-
-  VoxelFilter(const VoxelFilter&) = delete;
-  VoxelFilter& operator=(const VoxelFilter&) = delete;
+  explicit VoxelFilter(float size) : resolution_(size){};
```
Comment (pifon2a) [inline cartographer/sensor/voxel_filter.h]: clang-format?

File: cartographer/sensor/voxel_filter.cc [hunk 2]
```diff
@@ -76,31 +76,43 @@ PointCloud AdaptivelyVoxelFiltered(
...
+      results.push_back(point);
+    }
+  }
+  return results;
+}
 
-PointCloud VoxelFilter::Filter(const PointCloud& point_cloud) {
-  return FilterPointCloudUsingVoxels(point_cloud, &voxels_);
+VoxelFilter::KeyType VoxelFilter::IndexToKey(const Eigen::Array3i& index) {
+  KeyType b0(static_cast<uint32>(index[0]));
```
Comment (pifon2a) [inline cartographer/sensor/voxel_filter.cc]: nit: b_0, b_1,...

**Comment by gaschler** [comment]
@wally-the-cartographer merge

File: cartographer/sensor/voxel_filter.cc [hunk 3]
```diff
@@ -76,31 +76,43 @@ PointCloud AdaptivelyVoxelFiltered(
...
+  return results;
+}
 
-PointCloud VoxelFilter::Filter(const PointCloud& point_cloud) {
-  return FilterPointCloudUsingVoxels(point_cloud, &voxels_);
+VoxelFilter::KeyType VoxelFilter::IndexToKey(const Eigen::Array3i& index) {
+  KeyType k_0(static_cast<uint32>(index[0]));
+  KeyType k_1(static_cast<uint32>(index[1]));
+  KeyType k_2(static_cast<uint32>(index[2]));
+  return (k_0 << 2 * 8 * sizeof(int)) | (k_1 << 1 * 8 * sizeof(int)) | k_2;
```
Comment (cschuet) [inline cartographer/sensor/voxel_filter.cc]: why do you calculate the size as 8 * sizeof(int) and not sizeof(uint32)?

File: cartographer/sensor/voxel_filter.h [hunk 4]
```diff
@@ -43,7 +45,14 @@ class VoxelFilter {
   TimedPointCloud Filter(const TimedPointCloud& timed_point_cloud);
 
  private:
-  mapping::HybridGridBase<uint8> voxels_;
+  using KeyType = std::bitset<3 * 8 * sizeof(int)>;
```
Comment (cschuet) [inline cartographer/sensor/voxel_filter.h]: Same here. Why 8 * sizeof(int) instead of sizeof(uint32)? There are platforms where sizeof(int) != 4.

**Comment by gaschler** [comment]
@wally-the-cartographer merge

**Comment by wally-the-cartographer** [comment]
Merge requested by authorized user gaschler. Merge queue now has a length of 1.

**Pull request merged**